processor/src/hasher/tests.rs

```diff
@@ -0,0 +1,361 @@
+
+    let path1 = tree.get_path(1, 1).unwrap();
+    let new_leaf1 = init_leaf(4);
+    hasher.update_merkle_root(leaves[1], new_leaf1, &path1, Felt::ONE);
+    tree.update_leaf(1, new_leaf1).unwrap();
+
+    // build the trace
+    let trace = build_trace(hasher, 32);
+
+    // make sure the trace is correct
+    check_row_addr_trace(&trace);
+    check_selector_trace(&trace, 0, MR_UPDATE_OLD, RETURN_HASH);
+    check_selector_trace(&trace, 8, MR_UPDATE_NEW, RETURN_HASH);
+    check_selector_trace(&trace, 16, MR_UPDATE_OLD, RETURN_HASH);
+    check_selector_trace(&trace, 24, MR_UPDATE_NEW, RETURN_HASH);
+    check_hasher_state_trace(&trace, 0, hasher_merge_state(leaves[0], path0[0]));
+    check_hasher_state_trace(&trace, 8, hasher_merge_state(new_leaf0, path0[0]));
+    check_hasher_state_trace(&trace, 16, hasher_merge_state(path1[0], leaves[1]));
+    check_hasher_state_trace(&trace, 24, hasher_merge_state(path1[0], new_leaf1));
+    let node_idx_column = trace.last().unwrap();
+    assert_eq!(&node_idx_column[..16], &[Felt::ZERO; 16]);
+    assert_eq!(node_idx_column[16], Felt::ONE);
+    assert_eq!(&node_idx_column[17..24], &[Felt::ZERO; 7]);
+    assert_eq!(node_idx_column[24], Felt::ONE);
+    assert_eq!(&node_idx_column[25..], &[Felt::ZERO; 7]);
+
+    // --- Merkle tree with 8 leaves ------------------------------------------
+
+    // build a Merkle tree
+    let leaves = inti_leaves(&[1, 2, 3, 4, 5, 6, 7, 8]);
+    let mut tree = AdviceSet::new_merkle_tree(leaves.to_vec()).unwrap();
+
+    // initialize the hasher
+    let mut hasher = Hasher::new();
+
+    let path3 = tree.get_path(3, 3).unwrap();
+    let new_leaf3 = init_leaf(23);
+    hasher.update_merkle_root(leaves[3], new_leaf3, &path3, Felt::new(3));
+    tree.update_leaf(3, new_leaf3).unwrap();
+
+    let path6 = tree.get_path(3, 6).unwrap();
+    let new_leaf6 = init_leaf(25);
+    hasher.update_merkle_root(leaves[6], new_leaf6, &path6, Felt::new(6));
+    tree.update_leaf(6, new_leaf6).unwrap();
+
+    // update leaf 3 again
+    let path3_2 = tree.get_path(3, 3).unwrap();
+    let new_leaf3_2 = init_leaf(27);
+    hasher.update_merkle_root(new_leaf3, new_leaf3_2, &path3_2, Felt::new(3));
+    tree.update_leaf(3, new_leaf3_2).unwrap();
+    assert_ne!(path3, path3_2);
+
+    // build and check the trace for validity
+    let trace = build_trace(hasher, 144);
+    check_merkle_path(&trace, 0, leaves[3], &path3, 3, MR_UPDATE_OLD);
+    check_merkle_path(&trace, 24, new_leaf3, &path3, 3, MR_UPDATE_NEW);
+    check_merkle_path(&trace, 48, leaves[6], &path6, 6, MR_UPDATE_OLD);
+    check_merkle_path(&trace, 72, new_leaf6, &path6, 6, MR_UPDATE_NEW);
+    check_merkle_path(&trace, 96, new_leaf3, &path3_2, 3, MR_UPDATE_OLD);
+    check_merkle_path(&trace, 120, new_leaf3_2, &path3_2, 3, MR_UPDATE_NEW);
+}
+
+fn check_merkle_path(
+    trace: &[Vec<Felt>],
+    row_idx: usize,
+    leaf: Word,
+    path: &[Word],
+    node_index: u64,
+    init_selectors: Selectors,
+) {
```

Comment:
nit: I think it would make sense to have this down below with the other helper functions. I guess it's only used for the Merkle tree tests, but that's true of the init leaf/leaves functions too